dask/dataframe/core.py

```diff
@@ -6003,25 +5979,32 @@ def map_partitions(
                 divisions = [None] * (dfs[0].npartitions + 1)
 
     if has_keyword(func, "partition_info"):
-        dsk = dict(dsk)
-
-        for k, v in dsk.items():
-            subgraph = v[0]
-            number = k[-1]
-            assert isinstance(number, int)
-            info = {"number": number, "division": divisions[number]}
-            # Replace the __dummy__ keyword argument with `info`
-            subgraph_dsk = copy.copy(subgraph.dsk)
-            [(key, task)] = subgraph_dsk.items()
-            subgraph_dsk[key] = subs(task, {"__dummy__": info})
-            dsk[k] = (
-                SubgraphCallable(
-                    dsk=subgraph_dsk,
-                    outkey=subgraph.outkey,
-                    inkeys=subgraph.inkeys,
-                    name=f"{subgraph.name}-info-{tokenize(info)}",
-                ),
-            ) + v[1:]
+        partition_info = {
+            (i,): {"number": i, "division": division}
+            for i, division in enumerate(divisions[:-1])
+        }
+
+        args2.insert(0, BlockwiseDepDict(partition_info))
```

Comment:
This is an interesting application of `BlockwiseDep`. I don't think it's quite what was envisioned for the helper class, but it doesn't seem obviously wrong, and the win in this case is pretty clear. I wonder if it would be worth renaming that class or improving the docs, since it is not in this case strictly used for dependencies?

As far as I can tell, this is very similar to what @bmerry was trying in #7686 (though that one seems lazier, which may have some benefits). Unfortunately, that PR seems to have not gotten much attention, though it should have,

Reply:
Yup, the lack of laziness here was just laziness on my part :). This could be its own `BlockwiseDep` subclass that generates the info at construction time. Though since `divisions` would have to be embedded anyway, I felt like the savings from lazily generating `i` just weren't worth the effort.

I agree that this is a perhaps-unintended but very useful application of `BlockwiseDep`. I have a feeling that this is a pattern we'll want to reuse in many places. This is why I felt that https://github.com/dask/dask/issues/7513 would be powerful: it gives you an easy interface for passing in auxiliary data that needs to be broadcast like everything else, but is literal instead of delayed. I have a feeling that with a few basic constructs and the ability to pass in arrays/lists/iterables wrapped in a marker class, you could compose pretty much anything you need without defining new `BlockwiseDep` subclasses.

For example, I'd want to use this for https://github.com/dask/dask/issues/8445, in order to pass in the filenames here: https://github.com/dask/dask/blob/a5aecac8313fea30c5503f534c71f325b1775b9c/dask/dataframe/io/parquet/core.py#L738-L770

And yes, this seems pretty much exactly like what @bmerry was trying! Including the fact that you added support for passing a `BlockwiseDep` instance into `array.blockwise.blockwise`, which I also did in `dataframe.core.partitionwise_graph`, which to me is an equivalent function.

> I wonder if it would be worth renaming that class or improving the docs, since it is not in this case strictly used for dependencies?

Maybe `BlockwiseLiteral`? `BlockwiseArg`? I also perhaps like the idea of users not having to interact with it at all in basic cases—what if `map_partitions`/`map_blocks` automatically supported sequences or array-likes? (Just judging this on whether shapes align might be too much though; I could see wanting some explicit wrapping to clarify whether it's a literal or a broadcastable.) That plus `block_info` might get you everything you'd need as a user.